# Data Wrangling [4.4]

## Lesson Exercise

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# Assigning Path
path = r'/Users/elia/Desktop/DATA ANALYST/CAREER FOUNDRY/B - Data Analytics Immersion/4 - PYTHON FOR ANALYSTS/09-2025 Instacart Basket Analysis'

In [3]:
# Import Dataframe (Orders.csv)
df_ords = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
# Import Dataframe (Products.csv)
df_prods = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Dropping eval_set column from orders.csv
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [6]:
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [7]:
df_ords_2

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [8]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [9]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [10]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [11]:
df_ords

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
1         2398795        1    prior             2                   3   
2          473747        1    prior             3                   3   
3         2254736        1    prior             4                   4   
4          431534        1    prior             5                   4   
...           ...      ...      ...           ...                 ...   
3421078   2266710   206209    prior            10                   5   
3421079   1854736   206209    prior            11                   4   
3421080    626363   206209    prior            12                   1   
3421081   2977660   206209    prior            13                   1   
3421082    272231   206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [12]:
df_ords.head(25)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   1187899        1    train            11                   4   
11   2168274        2    prior             1                   2   
12   1501582        2    prior             2                   5   
13   1901567        2    prior             3                   1   
14    738281        2    prior             4                   2   
15   1673511        2    prior             5                   3   
16   1199898        2    prior             6                   2   
17   3194192        2    prior             7                   2   
18    788338        2    prior             8                   1   
19   1718559        2    prior             9                   2   
20   1447487        2    prior            10                   1   
21   1402090        2    prior            11                   1   
22   3186735        2    prior            12                   1   
23   3268552        2    prior            13                   4   
24    839880        2    prior            14                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0  
20                 11                     6.0  
21                 10                    30.0  
22                  9                    28.0  
23                 11                    30.0  
24                 10                    13.0

In [15]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [16]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [17]:
df_ords['order_id'].dtype

dtype('O')

In [18]:
# Import Dataframe (Departments.csv)
df_dep = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'departments.csv'), index_col = False)

In [19]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [20]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [21]:
df_dep_t = df_dep.T

In [22]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [24]:
# Take the first row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [25]:
new_header

0    department
Name: department_id, dtype: object

In [26]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [27]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [28]:
df_dep_t_new.columns = new_header # set the header row as the df header

In [29]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [30]:
data_dict = df_dep_t_new.to_dict('index')

In [31]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [32]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [33]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [34]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [35]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [36]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [37]:
df_prods.head(25)

product_id                                       product_name  aisle_id  \
0            1                         Chocolate Sandwich Cookies        61   
1            2                                   All-Seasons Salt       104   
2            3               Robust Golden Unsweetened Oolong Tea        94   
3            4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4            5                          Green Chile Anytime Sauce         5   
5            6                                       Dry Nose Oil        11   
6            7                     Pure Coconut Water With Orange        98   
7            8                  Cut Russet Potatoes Steam N' Mash       116   
8            9                  Light Strawberry Blueberry Yogurt       120   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
11          12                         Chocolate Fudge Layer Cake       119   
12          13                                  Saline Nasal Mist        11   
13          14                     Fresh Scent Dishwasher Cleaner        74   
14          15                           Overnight Diapers Size 6        56   
15          16                      Mint Chocolate Flavored Syrup       103   
16          17                                  Rendered Duck Fat        35   
17          18                Pizza for One Suprema  Frozen Pizza        79   
18          19   Gluten Free Quinoa Three Cheese & Mushroom Blend        63   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
20          21                   Small & Medium Dental Dog Treats        40   
21          22                  Fresh Breath Oral Rinse Mild Mint        20   
22          23                             Organic Turkey Burgers        49   
23          24  Tri-Vi-Sol® Vitamins A-C-and D Supplement Drop...        47   
24          25            Salted Caramel Lean Protein & Fiber Bar         3   

    department_id  prices  
0              19     5.8  
1              13     9.3  
2               7     4.5  
3               1    10.5  
4              13     4.3  
5              11     2.6  
6               7     4.4  
7               1     1.1  
8              16     7.0  
9               7     8.4  
10              7     2.8  
11              1     9.4  
12             11     1.1  
13             17     6.5  
14             18    11.2  
15             19     5.2  
16             12    17.1  
17              1    10.6  
18              9    12.0  
19              7     6.0  
20              8    10.0  
21             11     9.5  
22             12     8.2  
23             11     9.1  
24             19     1.9

In [38]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [39]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [40]:
df_snacks.head(25)

product_id                                       product_name  aisle_id  \
0             1                         Chocolate Sandwich Cookies        61   
15           16                      Mint Chocolate Flavored Syrup       103   
24           25            Salted Caramel Lean Protein & Fiber Bar         3   
31           32                      Nacho Cheese White Bean Chips       107   
40           41        Organic Sourdough Einkorn Crackers Rosemary        78   
55           56                         Healthy Pop Butter Popcorn        23   
58           59               Medium Taqueria Style Chipotle Salsa        50   
76           77                  Coconut Chocolate Chip Energy Bar         3   
77           78                  Nutter Butter Cookie Bites Go-Pak        61   
92           93                        Uncured Cracked Pepper Beef        23   
101         102                              Danish Butter Cookies        61   
102         103                              Sprouted Kale Cracker        78   
114         115                  Scooby-Doo! Fruit Flavored Snacks        50   
123         124                            Sun Cups Dark Chocolate        45   
134         135                   Dark Chocolate Ice Cream Topping       103   
135         136                                             Simple        45   
139         140                        Chips Onion Chipotle Garlic       107   
140         141  Restaurant Style Organic Chia & Quinoa Tortill...       107   
144         145                         Traditional Water Crackers        78   
155         156                     Extra Spearmint Sugar-Free Gum        46   
163         164                Peanut Butter Cacao Quinoa Clusters         3   
165         166                   Garlic Parmesan Pita Bread Chips       107   
171         172     Gluten Free All Natural Chocolate Chip Cookies        61   
179         180            Simply Beyond Black Bean Tortilla Chips       107   
182         183                    Cookie Dough Performance Energy         3   

     department_id  prices  
0               19     5.8  
15              19     5.2  
24              19     1.9  
31              19     4.9  
40              19     6.5  
55              19     4.4  
58              19     6.8  
76              19     6.4  
77              19     3.0  
92              19     2.4  
101             19     5.5  
102             19     6.4  
114             19     2.7  
123             19     3.0  
134             19     2.4  
135             19     5.0  
139             19     6.9  
140             19     2.9  
144             19     1.9  
155             19     3.5  
163             19     5.2  
165             19     5.2  
171             19     4.8  
179             19     3.0  
182             19     4.1

In [41]:
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [42]:
df_snacks_2

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [43]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [44]:
df_snacks_3

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# Data Wrangling [4.4]

## Task

#### Step 2 - Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [45]:
# First i will check some informations of the dataframe with Head Function.
df_ords.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329        1    prior             1                   2   
1  2398795        1    prior             2                   3   
2   473747        1    prior             3                   3   
3  2254736        1    prior             4                   4   
4   431534        1    prior             5                   4   
5  3367565        1    prior             6                   2   
6   550135        1    prior             7                   1   
7  3108588        1    prior             8                   1   
8  2295261        1    prior             9                   1   
9  2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0  
5                  7                    19.0  
6                  9                    20.0  
7                 14                    14.0  
8                 16                     0.0  
9                  8                    30.0

In [46]:
# Then i will check the type of data in each column
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(4), object(2)
memory usage: 182.7+ MB


In [47]:
# Then i will change order number to string, as i think there isn't any useful statistics i can get from just a order number.
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [48]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            object 
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


#### Step 3 - Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [49]:
# I do think the variable 'order_hour_of_day' is a bit confusing, therefore i will prefer to change it to 'hour_of_order'
df_ords_renamed = df_ords.rename(columns = {'order_hour_of_day' : 'hour_of_order'})

In [50]:
df_ords_renamed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            object 
 4   orders_day_of_week      int64  
 5   hour_of_order           int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


#### Step 4 - Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [51]:
# Using the value_counts command will give me at the top the highest number of orders in the corrisponding hour.
df_ords_renamed['hour_of_order'].value_counts()

hour_of_order
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

##### The busiest time of the day for placing orders is at around 10am.

#### Step 5 - Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [52]:
# I already have a dictionary created previously, so all i need to do is to call that id within that dictionary, using the print command.
print(data_dict.get('4'))

{'department': 'produce'}


#### Step 6 - The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [53]:
# With a boolean is very quick and straightforward to get this information, always relying on the data dictionary previously created.
df_breakfast =  df_prods[df_prods['department_id']==14]

In [54]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

#### Step 7 - They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [55]:
# With the isin command i can use a list and filter to the departments i do really need.
df_dinner_parties =  df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [56]:
df_dinner_parties.head(15)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
43          44                        Sparkling Raspberry Seltzer       115   
48          49      Vegetarian Grain Meat Sausages Italian - 4 CT        14   
51          52                                Mirabelle Brut Rose       134   
52          53              Whole Leaf Pure Aloe With Lemon Juice        98   
61          62            Premium Deli Oven Roasted Turkey Breast        96   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5  
43              7    12.2  
48             20    10.1  
51              5    14.4  
52              7     4.7  
61             20    14.6

#### Step 8 - It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [57]:
# Shape function is one of the basics in pandas and this is what is going to give me the information i need.
df_dinner_parties.shape

(7650, 5)

##### The last dataframe i've created has 7650 rows total.

#### Step 9 - Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [58]:
# This is very easy, i just need a boolean to give me back onlu the user id 1.
df_ords[df_ords['user_id'] == 1]

order_id  user_id eval_set order_number  orders_day_of_week  \
0   2539329        1    prior            1                   2   
1   2398795        1    prior            2                   3   
2    473747        1    prior            3                   3   
3   2254736        1    prior            4                   4   
4    431534        1    prior            5                   4   
5   3367565        1    prior            6                   2   
6    550135        1    prior            7                   1   
7   3108588        1    prior            8                   1   
8   2295261        1    prior            9                   1   
9   2550362        1    prior           10                   4   
10  1187899        1    train           11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

#### Step 10 - You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [59]:
# First i will check roughly the rows of customer_id 1,
# then i will create a subset just for that and then use the describe function to get the informations i need.
user_01_stats = df_ords[df_ords['user_id'] == 1]

In [60]:
user_01_stats.describe()

user_id  orders_day_of_week  order_hour_of_day  days_since_prior_order
count     11.0           11.000000          11.000000               10.000000
mean       1.0            2.636364          10.090909               19.000000
std        0.0            1.286291           3.477198                9.030811
min        1.0            1.000000           7.000000                0.000000
25%        1.0            1.500000           7.500000               14.250000
50%        1.0            3.000000           8.000000               19.500000
75%        1.0            4.000000          13.000000               26.250000
max        1.0            4.000000          16.000000               30.000000

#### Step 11 - Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [62]:
df_ords.shape

(3421083, 7)

In [64]:
df_ords.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'orders_wrangled.csv'))

#### Step 12 - Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [65]:
df_dep_t_new.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'departments_wrangled.csv'))